쥬스메이커 [STEP 3] Gundy, 준호

Gundy93 opened the pull request:
@junbangg
안녕하세요 알라딘! 🏜
추석은 잘 마무리하고 계신가요? 🌕
즐거운 한가위였기를 바랍니다. 😄
세 번째 PR도 잘 부탁드립니다. 🙌
이번 피드백도 감사합니다! 

#### 함수별 기능
- `updateFruitStock`: 각 과일의 레이블을 수량에 맞게 바꾸고 Stepper와 동기화 시키는 메서드, 과일이 목록에 없다면 레이블의 문구를 `FruitStoreError.notExist`로 바꾸고 Stepper를 비활성화 시킨다.
- `touchUpStepper`: 각 스텝퍼별로 `changeStock` 메서드를 호출해 재고의 변화를 과일 레이블에 적용시키는 메서드.
- `changeStock`: 스텝퍼의 값과 재고의 차를 변화량으로 반환하는 메서드.
- `showAddFruitsAlert`: 재고 목록에 없는 과일을 목록에 추가할 것인지 얼럿을 띄우는 메서드.
- `addFruit`: Juice의 recipe의 과일 중 재고목록에 없는 과일을 추가하는 메서드.
-  `fitText`: 화면이 작아서 쥬스 주문 버튼의 레이블 글자가 생략될 시 딱 맞춰주는 메서드.

### 코드를 작성할 때 고민되었던 점
- 화면 사이의 데이터 공유를 위해 Delegate 패턴을 활용했습니다. 
- FruitStore 타입의 fruitStore 프로퍼티를 요구하는 FruitStoreDelegate 프로토콜을 추가한 뒤, ViewController에서 채택했습니다.
- StockEditViewController에 FruitStoreDelegate 타입의 옵셔널 변수 delegate를 약한참조로 추가했습니다.
- StockEditViewController를 present 하기 전에 delegate를 self(ViewController)로 초기화했습니다.
- delegate를 사용해서 ViewController 프로퍼티 fruitStore의 currentStock, changeStock 메서드를 호출했습니다.
~~~swift
// FruitStoreDelegate 프로토콜 추가
protocol FruitStoreDelegate: AnyObject {
    var fruitStore: FruitStore { get set }
}


// ViewController에서 FruitStoreDelegate 프로토콜 채택
class ViewController: UIViewController, FruitStoreDelegate {

    // StockEditViewController의 delegate를 ViewController로 초기화
    func presentStockEditView() {
        guard let stockEditViewController = self.storyboard?.instantiateViewController(withIdentifier: "StockEditViewController") as? StockEditViewController else {
            return
        }
        stockEditViewController.delegate = 

// FruitStoreDelegate 프로토콜 타입의 delegate 추가 
class StockEditViewController: UIViewController {
    
    weak var delegate: FruitStoreDelegate?
    
    // delegate를 사용해서 currentStock, changeStock 메서드 호출
    func changeStock(of fruit: Fruit, by stockStepper: UIStepper) -> Int? {
        guard let currentStock = try? delegate?.fruitStore.currentStock(of: fruit) else {
            return nil
        }
        let stockAmountToChange: Int = Int(stockStepper.value) - currentStock
        delegate?.fruitStore.changeStock(of: fruit, by: stockAmountToChange)
        return currentStock + stockAmountToChange
    }
~~~
- 싱글턴을 사용하지 않은 이유는 싱글턴은 결국 `share`에 그 목적이 있다고 생각하기 때문입니다. "JuiceMaker라는 클래스는 FruitStore를 소유하고 있다"는 STEP 1의 요구조건과는 거리가 있다고 느껴져 싱글턴을 사용하지 않았습니다. FruitStore를 직접 소유한다는 점에서 델리게이트 패턴을 적용하는 것이 더욱 적절하다고 생각했습니다.
- 첫 화면과 겹치는 과일 이미지 및 재고 레이블은 첫 화면의 오토레이아웃 제약사항을 따랐습니다. 그 편이 사용자 입장에서 볼 때 더 자연스럽게 느껴질 것이라고 생각했습니다. Stepper의 경우는 기존 오토레이아웃 조건에서의 규칙을 그대로 적용하되, 사이즈 조절이 되지 않는 특성상 가운데 정렬을 우선적으로 따르게 설정하였습니다.
- 기존에는 재고수정 페이지로 화면전환을 할 때, 재고 수정 뷰 컨트롤러를 프레젠트 하였습니다. 하지만 네비게이션 바를 구현해야했기 때문에 `let navigationController = UINavigationController(rootViewController: stockEditViewController)` 선언부를 추가하고 네비게이션 컨트롤러를 프레젠트 하는 것으로 변경하였습니다.
- 재고를 수정하는 것이 어느 순간에 이루어지는 것이 적절한지에 대한 고민을 했습니다. 스텝퍼를 누를때마다 재고가 실시간으로 변경되어야 하는 것인지, 장바구니에 물건을 담기전 수량만 변경하는 것처럼 생각해 최종적으로 무언가 액션이 추가됐을 때 재고가 변경되어야 하는 것인지 고민을 했습니다. 결국 상의한 결과 레이아웃에 컴포넌트를 추가하지 않고 기능을 만족시키기 위해 즉각적으로 재고 수량이 변경되는 것으로 선택하였습니다.
- 버튼 타이틀 레이블의 글자가 길어지더라도 무관하게 자동으로 간격에 맞춰주기 위해 `fitText` 메서드를 만들었습니다. 

### 조언을 얻고 싶은 부분
- 현재 저희는 스토리보드의 인스펙터 탭을 통한 기능과 코드를 통한 기능을 함께 사용하고 있습니다. 때문에 해당 사항은 코드만 봐서는 구현되지않았는지 생각이 들 수도 있을 것 같습니다. 가령 저희는 스텝퍼의 미니멈밸류를 0으로 하고, 스텝퍼와 재고를 동기화시켰기때문에 스텝퍼를 통한 조작에서 재고는 0보다 작아질 수 없습니다. 하지만 이는 코드에서는 알 수 없는 부분이고, 혹시 누군가가 스토리보드를 통해 미니멈밸류를 수정한다면 발견하지 못하고 사용될 가능성도 있다고 생각합니다. 이같은 경우 코드상에서 더욱 분명히 하기 위해 `viewDidLoad` 메서드 등에서 `stepper.minimumValue = 0` 등의 코드를 작성하는 것이 좋을까요? 스토리보드를 건들지않는한 `stepper.minimumValue = 0`는 쓰이지 않는 코드가 됩니다. 이럴 때 어떤 선택이 더 좋을지 조언을 구하고 싶습니다.
- 오류처리를 위해 try를 사용하는 구간이 생기면 적절한 처리를 위해 고민을 했습니다. 하지만 `StockEditViewController`의 `changeStock` 메서드 같은 경우는 그저 nil을 반환할 뿐입니다. 이는 저희가 `updateFruitStock` 메서드를 통해 해당 오류가 발생할 원천인 스텝퍼를 비활성화했기 때문에 nil을 받는 경우가 절대 없으리라는 생각이 듭니다. 이런 경우 '아마도 절대 사용되지 않을 오류처리 코드'가 될지라도 '만에하나 나의 의도대로 스텝퍼가 비활성화되지않을 수 있으니까' nil을 반환하는 것보다도 더 적극적인 오류처리를 하는 것이 좋을지 궁금합니다.
- updateFruitStock 메서드는 Label의 text를 수정하고 Stepper의 value와 isEnabled를 수정하는 역할을 합니다. 한 메서드에서 한번에 많은 기능을 처리하는 데 어떤 방향으로 기능 분리를 하면 좋을까요? Label과 Stepper를 기준으로 분리를 해야할 지, if 블록 단위로 분리를 해야할 지... 조언을 얻고 싶습니다.
~~~swift
func updateFruitStock() {
    if let strawberryStock = try? delegate?.fruitStore.currentStock(of: .strawberry) {
        strawberryStockLabel.text = "\(strawberryStock)"
        straberryStockStepper.value = Double(strawberryStock)
    } else {
        strawberryStockLabel.text = FruitStoreError.notExist
        straberryStockStepper.isEnabled = false
    }
    ...
}
~~~
- `import Foundation`의 삭제와 같은 작업만으로 커밋을 진행할 때 메시지 서브젝트를 무엇으로 해야할 지 고민입니다. 우선 chore가 가장 적합한 것 같아서 그렇게 적용했는데, 어떤 카르마 스타일을 따라야 할까요?

junho15 committed:
refactor: 네비게이션 바 추가, 화면 전환 기능 수정

Gundy93 committed:
feat: 화면 진입시 과일의 현재 재고 수량을 표시하는 기능 추가, FruitStoreDelegate 프로토콜 추가

junho15 committed:
feat: Stepper를 이용해 재고 label 수정하는 touchUpStepper 메서드 추가

Gundy93 committed:
feat: Stepper를 통해 입력되는 값을 통해 재고를 변경시켜주는 changeStock 메서드 추가 및 기존 touchUpStepper 메서드 기능 분리

junho15 committed:
refactor: 오토레이아웃 적용

Gundy93 committed:
feat: FruitStoreError의 notInInventoryFruitList 케이스일 때 목록에 과일을 추가하는 addFruit 메서드 추가

junho15 committed:
chore: Foundation import 제거

Gundy93 committed:
feat: 작은 화면일 시 쥬스 주문 버튼의 폰트 사이즈를 딱 맞춰주는 fitText 메서드 추가

junbangg commented on JuiceMaker/JuiceMaker/Controller/StockEditViewController.swift:
```diff
@@ -8,7 +8,100 @@
+    @IBOutlet weak var bananaStockStepper: UIStepper!
+    @IBOutlet weak var kiwiStockStepper: UIStepper!
+    @IBOutlet weak var pineappleStockStepper: UIStepper!
+    @IBOutlet weak var mangoStockStepper: UIStepper!
```
Iboutlet 들도 은닉화가 가능하지않을까요?!

Gundy93 replied:
와 이건 전혀 생각하지 못했습니다 뷰 컨트롤러 내부의 은닉화를 더 신경쓰겠습니다 😃

Gundy93 replied:
5194bc5 커밋에서 은닉화 처리하였습니다!

junbangg commented on JuiceMaker/JuiceMaker/Controller/StockEditViewController.swift:
```diff
@@ -8,7 +8,100 @@
+    
+    @IBAction func touchUpStepper(_ sender: UIStepper) {
+        switch sender {
+        case straberryStockStepper:
```
딸기 오타가 있네요!

Gundy93 replied:
분명 둘이서 오타검수도...거쳤는데 눈물이 납니다 😂

Gundy93 replied:
5194bc5 커밋에서 오타수정했습니다! 😃

junbangg commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -6,7 +6,7 @@
-class ViewController: UIViewController {
+class ViewController: UIViewController, FruitStoreDelegate {
     
     var fruitStore: FruitStore = .init(inventory: [.strawberry: 10, .banana:10, .kiwi: 10, .mango: 10])
     lazy var juiceMaker: JuiceMaker = .init(fruitStore: fruitStore)
```
은닉화가 가능할것 같습니다 ..!안하기로 논의가 된 걸까요??

junbangg replied:
이거 지금 다시 보니까 생각이 들었는데ㅎㅎ
`.init` 을 선택한 이유가 있으실까요?!

Gundy93 replied:
11번 줄에서 이니셜라이저를 호출하면서 `FruitStore()`가 아닌 `.init()`을 사용한 이유에 대해 물어보시는 건가요? 아니면 12번 줄에서 `JuiceMaker()`가 아닌 `.init`을 사용한 이유에 대해 물어보시는 걸까요...

Gundy93 replied:
12번 줄의 `juiceMaker`에 대한 은닉화는 5194bc5 커밋에서 완료하였습니다. 11번 줄의 `fruitStore`는 `StockEditViewController`에서 `delegate`에 접근할 때 사용해야 하기 때문에 internal로 남겨두었습니다.

junbangg commented on JuiceMaker/JuiceMaker/Controller/StockEditViewController.swift:
```diff
@@ -8,7 +8,100 @@
+    @IBOutlet weak var bananaStockStepper: UIStepper!
+    @IBOutlet weak var kiwiStockStepper: UIStepper!
+    @IBOutlet weak var pineappleStockStepper: UIStepper!
+    @IBOutlet weak var mangoStockStepper: UIStepper!
```
여기 은닉화를 할 수 있는 친구들이 있어보여요!

Gundy93 replied:
5194bc5 커밋에서 은닉화작업 완료했습니다!

junbangg commented on JuiceMaker/JuiceMaker/Controller/StockEditViewController.swift:
```diff
@@ -8,7 +8,100 @@
+        guard let currentStock = try? delegate?.fruitStore.currentStock(of: fruit) else {
+            return nil
+        }
+        let stockAmountToChange: Int = Int(stockStepper.value) - currentStock
```
`stockStepper` 의 `value` 도 접근하고 형변환도 해주고 있으니까 요럴때는 상수로 따로 빼주는건 어떨까요?

Gundy93 replied:
cd186a4 커밋에서 65번라인에 `let stockStepperValue: Int = Int(stockStepper.value)` 따로 상수로 빼준 다음 66번 라인에서 `let stockAmountToChange: Int = stockStepperValue - currentStock` 사용하는 것으로 변경하였습니다.

junbangg commented on JuiceMaker/JuiceMaker/Controller/StockEditViewController.swift:
```diff
@@ -8,7 +8,100 @@
+    }
+    
+    func updateFruitStock() {
+        if let strawberryStock = try? delegate?.fruitStore.currentStock(of: .strawberry) {
```
요 메서드에 대해서 질문을 주셨는데!
우선 전반적으로 중복되는 코드가 꽤 많아보여서 메서드로 분리할 수도 있을것 같고요!
지금 과일에 대해서 각각 하나씩 처리를 해주고 있기 때문에 메서드가 길어지고 하는 일이 많아보이는게 아닐까 싶네요 저는!
과일이 234324235개가 생기는 경우에는 이렇게하면 손이 좀 아플것 같은데 이걸 좀 더 유연한 방법으로 처리할 방법은 없을까요?🤔

Gundy93 replied:
cd186a4 커밋에서 메서드 분리 및 `Fruit` 열거형 케이스의 원시값을 추가해서 레이블과 스테퍼의 태그로 활용하였습니다. 이 경우 과일이 더 늘어나고 스테퍼와 레이블이 늘어나더라도 과일의 원시값에 맞는 태그만 사용한다면 기존 함수 변경 없이 손쉽게 수정할 수 있다고 생각했습니다. 😃

- 추가로 조언을 구하고 싶습니다. 한 객체와 관련된 컴포넌트가 여러종류일 때, 2개까지는 `tag`를 사용할 수 있을 것 같습니다. 양수와 음수로 구분하면 되니까요. 하지만 3개 이상부터는 해당 객체와 컴포넌트들의 튜플로 구성된 딕셔너리를 정의해서 사용하는 것 밖에는 그다지 방법이 떠오르지 않습니다. 혹시 여러 컴포넌트가 한 객체와 관련될 때 어떤 방법을 사용해볼 수 있을까요?

junbangg commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -68,4 +68,12 @@ class FruitStore {
         return currentStock
     }
+    
+    func addFruit(of juice: Juice) {
```
혹시 이 메서드 하는 일이 뭐죠??

"addFruit: Juice의 recipe의 과일 중 재고목록에 없는 과일을 추가하는 메서드."

요렇게 설명을 해주셨는데 `addFruit` 라는 메서드명으로 설명이 충분히 되는지 모르겠습니다🤔

Gundy93 replied:
cd186a4 커밋에서 해당 메서드 `addNewFruitsOf(_ fruitList: [Fruit])`로 네이밍 변경하였습니다. '재고목록에 없는 과일'을 추가하는 만큼 'new'라는 단어를 사용하면 '기존에 없던' 것을 추가하는 것으로 이해되리라고 생각했습니다.

junbangg commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -140,4 +160,11 @@ class ViewController: UIViewController {
         }
     }
+    
+    func fitText() {
```
`fitText` 라는 메서드명만 봐서는 이게 어떤 일을 하는 친구인지 알기 힘들것 같아요!

Gundy93 replied:
cd186a4 커밋에서 해당 메서드 `adjustsFontSizeOfButtonsToFitWidth`로 네이밍 변경하였습니다.

junbangg commented on JuiceMaker/JuiceMaker/Model/FruitStoreDelegate.swift:
```diff
@@ -0,0 +1,10 @@
+//  Created by Gundy, 준호
+//
+
+protocol FruitStoreDelegate: AnyObject {
```
`Delegate`을 잘 써주셨군요👍👍

junbangg replied:
`Delegate` 패턴, `Singleton`, `NotificationCenter` 등 여러가지 방법들이 있었을텐데 `Singleton` 은 고민을 하셨고 `NotificationCenter` 에 대해서도 혹시 고려를 해보셨나요? (`NotifcationCenter` 가 맞다는 말은 아닙니다:) )

junho15 replied:
여기서는 화면 사이의 데이터 공유가 필요했습니다. StockEditViewController에서 ViewController에 있는 fruitStore의 재고 확인(currentStock), 수정(changeStock) 메서드를 호출해야 했습니다. 알림을 보내주는 NotificationCenter보다 Delegate 패턴이 이 경우에 더 적절하다고 생각했습니다.

junbangg reviewed (changes requested):
수고하셨습니다!!!!

이것저것 공부 열심히 하셔서 적용한 느낌을 받았습니다 👍👍👍

우선 질문 주신 부분!
1. Storyboard vs 코드 질문
이건 정말 정하기 나름인것 같아요...
왜 정해야되는지도 정확히 말씀해주신것 같아요!

2. 오류처리 질문

"아마도 절대 사용되지 않을 오류처리 코드"
"아마도" 이면 오류처리 해줄것 같고!
"절대" 이면 안할것 같아요 개인적으로

단, "절대" 라는 것은 진짜 논리적으로 불가능하다! 인거겠죠??
그럼 굳이 오류처리 안할것 같긴해요

근데 "절대" 라는게 그렇게 흔하게 발생하는 일은 아닐것 같다는 생각도 드네요...
그런 경우에는 저는 `nil` 대신 오류처리가 좋다는 생각이긴 합니다:)
정답은 없는것 같아요ㅋㅋ요건 저도 많이 고민해요!

3. 이건 댓글 달아드렸어요!

그다음에는 제가 추가적으로 이야기 해드리고 싶은 부부은
1. 
지금 PR Description에 메서드들에 대한 설명들을 친절하게 해주시고 있는데(너무 감동👍👍)
요설명 없이 메서드명만 보더라도 설명해주신 부분들이 최대한 짐작이 갈 수 있으면 더 좋은 메서드명들이 탄생할것 같아요:)

2. 
그리고 지금 타입별로 메서드들이 굉장히 많은데 `extension` 을 이용해서 정리를 해보는것이 어떨까요??
예를 들어 저는 `private` 메서드, `internal` 메서드들을 `extension` 을 이용해서 구분하는걸 좋아해요:)
요건 함께 디스코드에서 이야기 해봐도 좋을것 같네요 

3. 전반적으로 은닉화 체킹을 한 번 더 해주세요 ~ 👍 제가 놓친게 있을 수도 있어서 ..

junho15 committed:
refactor: 뷰 컨트롤러 프로퍼티와 메서드 은닉화,  오타 수정

Gundy93 committed:
refactor: 메서드 기능 분리 및 리네이밍

- Juice 열거형의 원시값 추가하여 스토리 보드의 태그로 활용
	- 레이블은 양수, 스테퍼는 음수로 사용
- updateFruitStock 메서드의 레이블 변경 기능을 하는 updateFruitStockLabel 메서드 분리
- updateFruitStock 메서드의 스테퍼 상태 변경 기능을 하는 updateFruitStockStepper 메서드 분리
- 기존 addFruit 메서드 addNewFruitsOf(_ fruitList: [Fruit])로 네이밍 변경
- 기존 fitText 메서드 adjustsFontSizeOfButtonsToFitWidth로 네이밍 변경
